geth/transactions/transactor.go

```diff
@@ -0,0 +1,205 @@
+package transactions
+
+import (
+	"context"
+	"math/big"
+	"sync"
+	"time"
+
+	ethereum "github.com/ethereum/go-ethereum"
+	gethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/status-im/status-go/geth/account"
+	"github.com/status-im/status-go/geth/rpc"
+	"github.com/status-im/status-go/sign"
+)
+
+const (
+	// sendTxTimeout defines how many seconds to wait before returning result in sentTransaction().
+	sendTxTimeout  = 300 * time.Second
+	rpcCallTimeout = time.Minute
+
+	defaultGas = 90000
+)
+
+// Transactor validates, signs transactions.
+// It uses upstream to propagate transactions to the Ethereum network.
+type Transactor struct {
+	pendingSignRequests  *sign.PendingRequests
+	sender               ethereum.TransactionSender
+	pendingNonceProvider PendingNonceProvider
+	gasCalculator        GasCalculator
+	sendTxTimeout        time.Duration
+	rpcCallTimeout       time.Duration
+	networkID            uint64
+
+	addrLock   *AddrLocker
+	localNonce sync.Map
+	log        log.Logger
+}
+
+// NewTransactor returns a new Manager.
+func NewTransactor(signRequests *sign.PendingRequests) *Transactor {
+	return &Transactor{
+		pendingSignRequests: signRequests,
+		addrLock:            &AddrLocker{},
+		sendTxTimeout:       sendTxTimeout,
+		rpcCallTimeout:      rpcCallTimeout,
+		localNonce:          sync.Map{},
+		log:                 log.New("package", "status-go/geth/transactions.Manager"),
+	}
+}
+
+// SetNetworkID selects a correct network.
+func (t *Transactor) SetNetworkID(networkID uint64) {
+	t.networkID = networkID
+}
+
+// SetRPCClient an RPC client.
+func (t *Transactor) SetRPCClient(rpcClient *rpc.Client) {
+	rpcWrapper := newRPCWrapper(rpcClient)
+	t.sender = rpcWrapper
```

Comment:
what is the reason to split?

Reply:
to avoid many responsibilities through the same interface. there are clearly 3 groups of them:
- the one that sends transactions (potentially we can even make our `Transactor` comply to the same interface;
- the one that provides information about gas prices;
- the one that provides information about pending nonce.

When I grouped all three of these, I really struggle with naming them anything except, well, a manager, a wrapper or a helper. It isn't really a transactor, because our transactor doesn't comply.

The benefits I see:
- simple naming and clear responsibilities;
- the code inside `verifyAndPropagate` now has constructions like `sender.SendTransaction` and `gasCalculator.EstimateGasPrice` which are very readable IMO;
- more flexibility for potential testing if we want to test only a part of this interface.

The downside:
- more interfaces
- more fields

I think it worth the deal.

Reply:
not sure, usually, it is the same object that provides all of this methods

Reply:
maybe it is too java-ish, but still :)
I'm struggling with naming everything consistently here, so we don't end up with 2 different transactors like @adambabik pointed out.

Reply:
Since looks like this is the only part left.
I'd maybe leave it for another thought later.
These names are still better than `TxQueueManager`, IMO